### ISSUE TITLE
Support for generating prop types?

### ISSUE THREAD

**Issue by selbekk** [jaredpalmer/tsdx#418]
### Current Behavior

Right now, projects made with TSDX doesn't generate prop types automatically.

### Desired Behavior

I'd like to have an option (or a way to extend the config) so that I can generate prop types in the development version of the package I create.

### Suggested Solution

We could add a config option to add this babel plugin: 

https://www.npmjs.com/package/babel-plugin-typescript-to-proptypes

Or write something similar, idk 🤷‍♂️

### Who does this impact? Who is this for?

We're creating a component library in TypeScript, but many of our consumers don't use TypeScript. It would be great to add prop types as a fallback runtime verification for those that aren't using types. 

This translates well to many packages created with TSDX, I would assume.

**Comment by ambroseus** [comment]
@selbekk  you have such option: https://github.com/jaredpalmer/tsdx#babel

**Comment by selbekk** [comment]
🤔 I tried to generate a repro:

- bootstrap a new project with `tsdx create example` (with React)
- add some types to the generated `src/index.tsx` component
- install the `babel-plugin-typescript-to-proptypes`
- add a `.babelrc` with the following config

```json
{
  "plugins": ["babel-plugin-typescript-to-proptypes"]
}
```

And it didn't make a difference to the output at all. Am I missing an important step?

Here's a repro: https://github.com/selbekk/repro-tsdx-418

**Comment by ambroseus** [comment]
@selbekk  *thinking*  guess this plugin is not applicable with tsdx cause babel get already compiled sources after rollup (not original typescrpt) https://github.com/rollup/rollup-plugin-babel#why

I'll check it more deeply soon...

**Comment by ambroseus** [comment]
@selbekk  maybe you don't need tsdx for building react  components library
https://hackernoon.com/creating-a-library-of-react-components-using-create-react-app-without-ejecting-d182df690c6b

**Comment by selbekk** [comment]
TSDX is a great way to create packages, and my component library is just a bunch of packages, really. :) I don't see why I couldn't use TSDX for this.

**Comment by ambroseus** [comment]
yep, tsdx is great! :) but not a silver bullet.. but this is interesting challenge so I'll try to solve it :)

**Comment by ambroseus** [comment]
@selbekk  I have solution, but it's not so simple))
![Screenshot from 2020-01-08 20-38-22](https://user-images.githubusercontent.com/380645/72005772-08d5e300-3257-11ea-8fdc-0d0bb7a8ee40.png)

**Comment by ambroseus** [comment]
btw, I also prefer to use typescript only as type checker and type defs generator - not as transpiler and compiler - this is for babel which is first class transpiler :)

**Comment by selbekk** [comment]
I'm happy that works, and thanks for the incredible effort solving that (!), but switching out the entire compiler is a bit convoluted. Is there no way this could be built in to tsdx itself? as a rollup plugin or whatever?

**Comment by ambroseus** [comment]
as I see the only way is to extend rollup-plugin-typescript2 with custom typescript transformer
(like this one https://github.com/ambroseus/tsdx/pull/1/commits/f20dc6912d1d1cabba1c4daec2c68c84c47806e1)
but  as @jaredpalmer says this is not `tsdx` way..

**Comment by ambroseus** [comment]
@selbekk  found another solution: https://github.com/gcanti/prop-types-ts

**Comment by ambroseus** [comment]
...to many `but` :smile:

**Comment by ambroseus** [comment]
well.... 
> As for me, I think I will have to keep both TypeScript typings and manual PropTypes in sync for now, because I'm not using Babel and I don't really like a custom PropTypes implementation. Maybe, one sunny day, we will find some resources to actually write our own PropTypes generator in our company and will share it with everyone.

**Comment by agilgur5** [comment]
There is also `PropTypes.InferProps`, which is detailed well in https://dev.to/busypeoples/notes-on-typescript-inferring-react-proptypes-1g88

**Comment by acasazza** [comment]
Hey guys! It will be awesome! When will you be in the roadmap?

**Comment by agilgur5** [comment]
@acasazza please read the comments.
There isn't currently a viable approach, so there can't be a roadmap.

**Comment by acasazza** [comment]
@agilgur5, I tried babel-plugin-typescript-to-proptypes, and it seems the better solution. Can I make a tsdx template? What do you think?

**Comment by agilgur5** [comment]
@acasazza again, please read the comments before notifying everyone...

`babel-plugin-typescript-to-proptypes` is not compatible with TSDX.
It requires using `@babel/preset-typescript` to strip types whereas TSDX fully compiles the TypeScript with `rollup-plugin-typescript2`.

That was stated in above comments and why I said it's not viable.

**Comment by andriyor** [comment]
I think this comment can help you https://github.com/milesj/babel-plugin-typescript-to-proptypes/issues/8#issuecomment-604709978

**Comment by ambroseus** [comment]
@andriyor  thanks)
material-ui alredy use it https://github.com/mui-org/material-ui/pull/21497

**Comment by agilgur5** [comment]
Thought I'd mention this here for any readers, but I wrote upstream in rpt2 (where I now help out as a maintainer) in https://github.com/ezolenko/rollup-plugin-typescript2/issues/268#issuecomment-1151665344 that using Babel plugins on top of TS code _may_ be possible with a two-pass system.
First run through [`@babel/plugin-syntax-typescript`](https://babeljs.io/docs/en/babel-plugin-syntax-typescript) to allow Babel to understand TS but _not_ strip it. Add some TS specific plugins only, then you can pass it through rpt2, and then back to Babel to use things like `preset-env` etc.

It's very inefficient and hacky, but could be doable if someone wants to do so. Also code returned by Babel plugins is likely pure JS so it may complicate type-checking. Experimentation definitely required, but thought I might as well mention it.
